### ISSUE TITLE
UDP only: connection timed out; no servers could be reached

### ISSUE THREAD

**Issue by sebolabs** [NLnetLabs/unbound#707]
**Describe the bug**
Possibly me doing something wrong, but need your help!

The issue I encountered is that unbound does not reply to queries from remote networks.
I have a local network (192.168.0.0/24) and a remote AWS network (10.0.0.0/8).
They are integrated with the use of VPN and TGW.
All traffic between both networks is allowed, netcat/telnet are happy.

From any host in the local network UDP queries work fine.
However, when I try to query `unbound` from the remote AWS network it fails with:
```
root@ip-10-0-2-6:~# dig xxx.dummy.blah @192.168.0.158

; <<>> DiG 9.18.1-1ubuntu1.1-Ubuntu <<>> xxx.dummy.blah @192.168.0.158
;; global options: +cmd
;; connection timed out; no servers could be reached
``` 
If I only add `+tcp` flag it gets a successful response.

**To reproduce**
N/A

**Expected behavior**
It should respond to UDP queries

**System:**
 - Unbound version: `1.9.4-2ubuntu1.2`
 - OS: `Linux pi 5.4.0-1065-raspi #75-Ubuntu SMP PREEMPT Fri Jun 3 14:24:57 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux`
 - unbound -V: 
``` 
Version 1.9.4

Configure line: --build=aarch64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir=${prefix}/lib/aarch64-linux-gnu --libexecdir=${prefix}/lib/aarch64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --disable-rpath --with-pidfile=/run/unbound.pid --with-rootkey-file=/var/lib/unbound/root.key --with-libevent --with-pythonmodule --enable-subnet --enable-dnstap --enable-systemd --with-chroot-dir= --with-dnstap-socket-path=/run/dnstap.sock --libdir=/usr/lib --disable-flto
Linked libs: libevent 2.1.11-stable (it uses epoll), OpenSSL 1.1.1f  31 Mar 2020
Linked modules: dns64 python subnetcache respip validator iterator
```

**Additional information**
My unbound.conf
```
server:
  port: 53
  interface: 0.0.0.0

  verbosity: 4

  num-threads: 2
  outgoing-range: 512
  num-queries-per-thread: 1024
  msg-cache-size: 32m
  rrset-cache-size: 64m
  cache-max-ttl: 86400
  infra-host-ttl: 60
  infra-lame-ttl: 120

  access-control: 127.0.0.0/8 allow
  access-control: 192.168.0.0/24 allow
  access-control: 10.0.0.0/8 allow

  directory: "/etc/unbound"
  use-syslog: yes
  hide-version: yes

  so-rcvbuf: 2m
  so-sndbuf: 2m

  do-ip4: yes
  do-ip6: no
  do-udp: yes
  do-tcp: yes

  local-data: "xxx.dummy.blah A 192.168.0.158"

remote-control:
  control-enable: yes
  control-interface: 127.0.0.1
```

**Comment by wcawijngaards** [comment]
If you enable `log-queries: yes` unbound starts logging every individual query, to the log with one line. And that can tell you if queries make it to unbound. From your config, unbound is listening to the wildcard IP address and seems to allow the queries, and unbound is listening to UDP. You can also use `log-replies: yes`.

Is there a firewall rule, maybe on the pc that runs unbound, and you have it set to allow TCP but not UDP?

It could also be a routing problem, somehow for UDP but not TCP. Seems unlikely, but if so, specifying different interface lines could help. Explicitly specifying the interface, or interfaces on separate lines. Or use `interface-automatic: yes` that copies the address the packet was received on as the source of the answer.

Perhaps you have the network set up for TCP but not for UDP? I mean the integration method. That would explain how local network hosts can use UDP, but remote ones cannot because the integration of the networks does not allow their UDP traffic.

**Comment by sebolabs** [comment]
Here's some output after I enabled logging queries and replies:
**TCP (remote)**
```
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: comm point start listening 18 (30000 msec)
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: Reading tcp query of length 58
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] info: 10.0.2.6 xxx.dummy.blah. A IN
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: using localzone xxx.dummy.blah. transparent
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] info: 10.0.2.6 xxx.dummy.blah. A IN NOERROR 0.000000 1 62
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: comm point stop listening 18
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: comm point start listening 18 (30000 msec)
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: comm point stop listening 18
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: comm point start listening 18 (30000 msec)
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: tcp channel read side closed 18
Jul  4 15:20:57 localhost unbound[6415]: [1656948057] unbound[6415:1] debug: close fd 18
```
... all good

**UDP (remote)**
```
Jul  4 15:21:36 localhost unbound[6415]: [1656948096] unbound[6415:1] info: 10.0.2.6 xxx.dummy.blah. A IN
Jul  4 15:21:36 localhost unbound[6415]: [1656948096] unbound[6415:1] debug: using localzone xxx.dummy.blah. transparent
Jul  4 15:21:36 localhost unbound[6415]: [1656948096] unbound[6415:1] info: 10.0.2.6 xxx.dummy.blah. A IN NOERROR 0.000000 1 62
Jul  4 15:21:41 localhost unbound[6415]: [1656948101] unbound[6415:0] info: 10.0.2.6 xxx.dummy.blah. A IN
Jul  4 15:21:41 localhost unbound[6415]: [1656948101] unbound[6415:0] debug: using localzone xxx.dummy.blah. transparent
Jul  4 15:21:41 localhost unbound[6415]: [1656948101] unbound[6415:0] info: 10.0.2.6 xxx.dummy.blah. A IN NOERROR 0.000000 1 62
Jul  4 15:21:46 localhost unbound[6415]: [1656948106] unbound[6415:0] info: 10.0.2.6 xxx.dummy.blah. A IN
Jul  4 15:21:46 localhost unbound[6415]: [1656948106] unbound[6415:0] debug: using localzone xxx.dummy.blah. transparent
Jul  4 15:21:46 localhost unbound[6415]: [1656948106] unbound[6415:0] info: 10.0.2.6 xxx.dummy.blah. A IN NOERROR 0.000000 1 62
```
... no banana 😞 

Then after enabling the `interface-automatic` option, another remote UDP check:
```
Jul  4 15:29:46 localhost unbound[6620]: [1656948586] unbound[6620:0] info: receive_udp on interface: 15 192.168.0.158 192.168.0.158
Jul  4 15:29:46 localhost unbound[6620]: [1656948586] unbound[6620:0] info: 10.0.2.6 xxx.dummy.blah. A IN
Jul  4 15:29:46 localhost unbound[6620]: [1656948586] unbound[6620:0] debug: using localzone xxx.dummy.blah. transparent
Jul  4 15:29:46 localhost unbound[6620]: [1656948586] unbound[6620:0] info: 10.0.2.6 xxx.dummy.blah. A IN NOERROR 0.000000 1 62
Jul  4 15:29:46 localhost unbound[6620]: [1656948586] unbound[6620:0] info: send_udp over interface: 15 192.168.0.158 192.168.0.158
```

Success! It started working 🚀 
Thx!

**Comment by wcawijngaards** [comment]
Good to hear that it works, closing the issue.

**Comment by v-bryce** [comment]
After much research and head scratching, I came across this unbound.conf. Basically, I was getting "connection refused" errors while using dig to check DNS functionality from both the local machine (using its proper address - 192.168.1.5) and from a remote machine. After reading through and trying numerous "fixes", I replaced my unbound.conf entries with these.

It worked first time. The only change I made to this config was to add in the DNSSEC entries and remove the 10.x.x.x network.

Happy days :)
